lib/src/vector_math/matrix4.dart

```diff
@@ -704,27 +766,83 @@ class Matrix4 {
   }
 
   /// Multiply this by a translation from the left.
-  /// The translation can be specified with a  [Vector3], [Vector4], or x, y, z.
+  ///
+  /// The translation can be specified with a [Vector3], [Vector4], or x, y, z
+  /// as [double]s.
+  ///
+  /// If you know the argument types in a call site, prefer
+  /// [leftTranslateByDouble], [leftTranslateByVector3], or
+  /// [leftTranslateByVector4] for performance.
+  @pragma('wasm:prefer-inline')
+  @pragma('vm:prefer-inline')
   void leftTranslate(dynamic x, [double y = 0.0, double z = 0.0]) {
-    double tx;
-    double ty;
-    double tz;
-    final tw = x is Vector4 ? x.w : 1.0;
     if (x is Vector3) {
-      tx = x.x;
-      ty = x.y;
-      tz = x.z;
+      leftTranslateByVector3(x);
     } else if (x is Vector4) {
-      tx = x.x;
-      ty = x.y;
-      tz = x.z;
+      leftTranslateByVector4(x);
     } else if (x is double) {
-      tx = x;
-      ty = y;
-      tz = z;
+      leftTranslateByDouble(x, y, z);
     } else {
       throw UnimplementedError();
     }
+  }
+
+  /// Multiply this by a translation from the left.
+  void leftTranslateByDouble(double tx, double ty, double tz) {
+    // Column 1
+    _m4storage[0] += tx * _m4storage[3];
```

Comment:
Consider hoisting the `_m4storage[3]` into a variable as some compilers may not perform this optimization

Also consider moving the `_m4storage[15]` operation first.

(also in other cases)